### DISCUSSION TITLE
Another FSK question

### DISCUSSION THREAD

**Discussion by Scoff123** [dbuezas/esphome-cc1101#3] [Q&A]
Hi @dbuezas 

I came here after you pointed me in this direction from the HA community forum - thanks for all your efforts making this esphome-cc1101 available! :)

I have installed ESPHome on an ESP32 DEVKITV1 board
<img width="402" alt="image" src="https://user-images.githubusercontent.com/96984460/180401247-cbde2855-bed6-47d2-8d67-ce3188dad1aa.png">

and connected this to a D-SUN CC1101 433 board as follows:
CC1101  ESP32
VCC -> 3v3
GND -> GND
SCK -> D14
MOSI -> D13
GDO2 -> n/a
MISO -> D12
CSN -> D15
GDO0 ->D25

However, am I correct in thinking I don't need to use these specific Hardware SPI pins, as per your answer on [discussion #1](https://github.com/dbuezas/esphome-cc1101/discussions/1) "The way this component works is by putting the cc1101 in "dumb mode" and send/receive the information through a single pin (instead of using SPI)"

Will using these hardware SPI pins cause an issue in this scenario - should I use alternative pins instead? My ESPHome yaml code compiles ok and installs fine, but I haven't created any sensors within it using the raw 433 codes yet, as I just want to 'sniff' what I believe to be an FSK transmission first to ensure my hardware and basic code works and picks up the raw data. Currently not seeing any raw data in my ESPHome logs for this board.

**Comment by dbuezas** [comment]
Hey Scoff!
The custom component does use spi, but only to configure the cc1101, not to transmit the data. This means that the pins should be ok.
Regarding FSK, the trick is that there is one extra parameter apart from the base frequency: FSK works by changing the base frequency. This means that instead of turning the radio on and off to communicate, it jumps between one frequency and another one. The difference between those frequencies is the unknown parameter for you.

Ideally, you can find it out using a "software defined radio" (usually an RTL-SDR), but if you don't want to get one, the only option I see is to change that parameter by hand until you start receiving something.

To do that, I recommend you set up a slider or numeric imput for home assistant. Check the other discussion and also the "advanced" yaml in this repo.

Much luck!

(But srsly, buy an sdr, not expensive and very fun to play with)

**Reply by Scoff123** [reply]
Great, thanks for the pointers 👍 New to all things RF related, so trying to learn as much as possible to help achieve my goal! It does sound like it's time to pick up an SDR, as I do like tinkering and discovering things :) There may be local weather stations I can pick up as well as the device I want to pick up. 

I know that RTL_433 supports the FSK modulation of my device using protocol 43, but it doesn't look like the ESPHome version of RTL_433 supports this protocol yet. If I can use thr Windows port of RTL_433 with an SDR to get me started, I can then look to hopefully use this ESPHome-cc1101 setup. 

I'm off to find an SDR.......

**Reply by dbuezas** [reply]
Oh, then here's the decoder: https://github.com/merbanan/rtl_433/blob/6e55cb64741fc4aae87d34a205f530ffd94d18d0/src/devices/oil_watchman.c

I'm not sure what fsk pulse pcm is though.
I also don't understand the fact that there isn't any mention to both frequencies either.

Finally, you won't be able to use the remote_receiver component on this, as that one can only detect specific patterns (and not parse arbitrary numbers).
The good news is that the link above contains the actual decoding algorithm, so if you can code in C, you may be able to integrate it in EspHome anyway :) 💪

**Reply by dbuezas** [reply]
https://github.com/merbanan/rtl_433/blob/f8d98dfdc3bc11eb1af22da79b3cd5ac3637fb37/include/r_device.h#L36

FSK_PULSE_PCM                = 16, ///< FSK Modulation, Non-Return-to-Zero coding, Pulse = 1, No pulse = 0.

I think that's "usual" fsk

**Reply by dbuezas** [reply]
I looked at the code a bit and I can't find where the distance between both frequencies is defined. 
Maybe the rtl_433 library some how is able to detect FSK without knowing that in advance

**Reply by Scoff123** [reply]
Thank you again for the information, and taking the time to look into it! I'm a little over my head as I have no coding experience, just general IT and networking, but am keen to figure this out. As the necessary decoding algorithm is available in that link that you mentioned above, what would be required in order to make that available within ESPHome? I'm happy to research if i can get a little guidance as to what elements are required to bring it all together in ESPHome :)

**Reply by dbuezas** [reply]
It will be tricky, you need to:
1. Find the right parameters for the cc1101 to see the signal at all
2. Collect the signal in the esp32 and keep it in a circular buffer with the same format as rtl433 does
3. Call the decoder each time a new "bit" is received
4. Set up a custom component in EspHome
5. Publish the collected data within that custom sensor.

Alternatively, the cc1101 may be able to do the buffering and part of the decoding itself.

If you're not into coding, I think you'll have more fun and success connecting an rtl-sdr directly to HA. I haven't tried it myself, but here's a lead: https://community.home-assistant.io/t/home-assistant-add-on-rtl-433-with-mqtt-auto-discovery/260665.

Since you already wired the cc1101, you could still use it as a 433 gateway for switches and the like :)

**Reply by Scoff123** [reply]
Thanks for listing out the requirements! Although I am getting better with ESPHome, I think for now the SDR is the way forward like you suggest :) My RTL-SDR should arrive tomorrow :) Looking forward to playing around and getting my head round it all. I've found some beginner guides to using a RTL-SDR with various free software so that should help. Then I can see about picking up my watchman FSK device. The CC1101 and sonof rf bridge will no doubt serve a purpose once I'm more proficient with RF.....

**Reply by dbuezas** [reply]
My pleasure, wish you much fun with it!

**Reply by Scoff123** [reply]
I've been having fun with it so far! Started with picking up mainstream FM radio stations to check my SDR works, then running the windows version of RTL_433 to see what I could pick up :)

![image](https://user-images.githubusercontent.com/96984460/180971194-093712a0-b335-4aaf-9bbd-1818f99983ec.png)

Picking up what appears to be 2 of my neighbours Oil tank monitors, but not my own! I turned my transmitter off for a couple of hours to try to identify which one was my one, but both of these continued to send readings, so perhaps my transmitter isn't FSK after all. It's definitely 433.92 as this is printed on the case and the transmitter, so I think I will need to do some sniffing to try to identify my transmitter. My plug in Watchman Plus receiver works fine and receives the signal from my transmitter so I know it works, just need to identify the keying I believe..... enjoying playing with it so far though, thanks for the suggestion on the SDR :)

**Reply by dbuezas** [reply]
👏 👏👏
Here's an idea, download a spectrogram app and tune it around 433Mhz. Then wait until the signal comes and post the waterfall here. With that you may be able to find out if it is fsk and find more clues about what the device transmits
